src/biome/text/modules/heads/token_classification.py

```diff
@@ -216,29 +234,53 @@ def forward(  # type: ignore
 
         return output
 
+    def _decode_tags(self, k_tags: List[List[int]]) -> List[List[str]]:
+        """Decode predicted tags"""
+        return [
+            [vocabulary.label_for_index(self.backbone.vocab, idx) for idx in tags]
+            for tags, prob in k_tags
+        ]
+
+    def _decode_entities(
+        self, doc: Doc, tags: List[str], pre_tokenized: bool
+    ) -> List[Dict]:
+        """Decode predicted entities from tags"""
+        return offsets_from_tags(
+            doc, tags, self._label_encoding, only_token_spans=pre_tokenized
+        )
+
+    def _decode_tokens(self, doc: Doc) -> List[Dict]:
+        """Decode tokens"""
+        return [
+            {"text": token.text, "start": token.idx, "end": token.idx + len(token)}
+            for token in doc
+        ]
+
     def decode(self, output: TaskOutput) -> TaskOutput:
         # Te dims are: batch, k_tags, tags
         output.tags: List[List[List[str]]] = [
-            [
-                [vocabulary.label_for_index(self.backbone.vocab, idx) for idx in tags]
-                for tags, prob in k_tags
-            ]
-            for k_tags in output.viterbi_paths  # loop over batch
+            self._decode_tags(k_tags) for k_tags in output.viterbi_paths
         ]
-        del output.viterbi_paths
 
-        if isinstance(output.raw_text[0], str):
-            entities: List[List[List[Dict]]] = []
-            for raw_text, k_tags in zip(output.raw_text, output.tags):
-                doc = self.backbone.tokenizer.nlp(raw_text)
-                top_k_entities: List[List[Dict]] = [
-                    offsets_from_tags(doc, tags, self._label_encoding)
-                    for tags in k_tags
-                ]
-                entities.append(top_k_entities)
-            output.entities = entities
-
-        del output.raw_text
+        # fmt: off
+        output.entities, output.tokens = zip(*[
+            (
+                [self._decode_entities(doc, tags, pre_tokenized) for tags in k_tags],
+                self._decode_tokens(doc) if not pre_tokenized else None
+            )
+            for doc, pre_tokenized, k_tags in zip(*output.tokenization, output.tags)
+        ])
+        # fmt: on
```

Comment:
puf, i would prefer to go a bit slower here ... this will take me 10 min minimum to understand every time i look at it ... :)
```suggestion
        output.entities, output.tokens = [], []
        for doc, pre_tokenized, k_tags in zip(*output.tokenization, output.tags):
            output.entities.append([self._decode_entities(doc, tags, pre_tokenized) for tags in k_tags])
            output.tokens.append(self._decode_tokens(doc) if not pre_tokenized else None)
```
I did not test the suggestion, so maybe it includes a typo o a stupid bug somewhere ...

Reply:
Maybe it's time to learn python :-)